src/etlng/impl/Extraction.cpp

```diff
@@ -0,0 +1,227 @@
+            return fallback;
+        return maybe;
+    };
+
+    return {
+        .key = *key,  // trivially copyable
+        .keyRaw = std::move(*obj.mutable_key()),
+        .data = {obj.mutable_data()->begin(), obj.mutable_data()->end()},
+        .dataRaw = std::move(*obj.mutable_data()),
+        .successor = value_or(obj.successor(), uint256ToString(data::firstKey)),
+        .predecessor = value_or(obj.predecessor(), uint256ToString(data::lastKey)),
+        .type = extractModType(obj.mod_type()),
+    };
+}
+
+std::vector<model::Object>
+extractObjs(PBObjListType objects)
+{
+    namespace rg = std::ranges;
+    namespace vs = std::views;
+
+    // TODO: should be simplified with ranges::to<> when available
+    std::vector<model::Object> output;
+    output.reserve(objects.size());
+
+    rg::move(objects | vs::transform([](auto&& obj) { return extractObj(obj); }), std::back_inserter(output));
+    return output;
+}
+
+model::BookSuccessor
+extractSuccessor(PBBookSuccessorType successor)
+{
+    return {
+        .firstBook = successor.first_book(),
+        .bookBase = successor.book_base(),
+    };
+}
+
+std::optional<std::vector<model::BookSuccessor>>
+maybeExtractSuccessors(PBLedgerResponseType const& data)
+{
+    namespace rg = std::ranges;
+    namespace vs = std::views;
+
+    if (not data.object_neighbors_included())
+        return std::nullopt;
+
+    // TODO: should be simplified with ranges::to<> when available
+    std::vector<model::BookSuccessor> output;
+    output.reserve(data.book_successors_size());
+
+    rg::copy(
+        data.book_successors() | vs::transform([](auto&& obj) { return extractSuccessor(obj); }),
+        std::back_inserter(output)
+    );
+    return output;
+}
+
+auto
+Extractor::makeExtractor(uint32_t seq)
+{
+    return [seq](auto&& data) {
+        auto header = ::util::deserializeHeader(ripple::makeSlice(data.ledger_header()));
+
+        return std::make_optional<model::LedgerData>({
+            .transactions =
+                extractTxs(std::move(*data.mutable_transactions_list()->mutable_transactions()), header.seq),
+            .objects = extractObjs(std::move(*data.mutable_ledger_objects()->mutable_objects())),
+            .successors = maybeExtractSuccessors(data),
+            .edgeKeys = std::nullopt,
+            .header = header,
+            .rawHeader = std::move(*data.mutable_ledger_header()),
+            .seq = seq,
+        });
+    };
+}
+
+std::optional<model::LedgerData>
+Extractor::extractLedgerWithDiff(uint32_t seq)
+{
+    LOG(log_.debug()) << "Extracting DIFF " << seq;
+
+    auto [batch, time] = ::util::timed<std::chrono::duration<double>>([this, seq] {
+        return fetcher_->fetchDataAndDiff(seq).and_then(makeExtractor(seq));
+    });
+
+    LOG(log_.debug()) << "Extracted and Transformed diff for " << seq << " in " << time << "ms";
+
+    // can be nullopt. this means that either the server is stopping or another node took over ETL writing.
```

Comment:
I feel it still can fetch valid data from grpc when another node took over.

Reply:
This depends on the fetcher.. out of scope of this PR i'd say. Note that this stuff will still change more and fetcher itself will be updated later too. This comment basically migrated from previous logic